File: python/lexer.py
import string
from objects import Token
from error import TNLUnidentifiedToken, TNLSyntax

token_types = [
    'datatype',
    'operator',
    'punctuator',
    'number',
    'string',
    'endstatement',
    'assignment',
    'rawstring'
]

data_types = [
    'int',
    'str'
]


def tokenize(code: str) -> list[Token]:
    all_tokens = []
    current_string = ""
    current_number = ""

    if len(code) == 0:
        raise TNLSyntax("Error: Empty code string.")

    for i, char in enumerate(code):
        if char.isspace() or char == '\n':
            if current_string:
                if current_string in data_types:
                    token_type = token_types[0]
                else:
                    token_type = token_types[4]
                token_pos = (i - len(current_string), i)
                all_tokens.append(Token(token_type, current_string, token_pos))
                current_string = ""
            elif current_number:
                token_type = token_types[3]
                token_pos = (i - len(current_number), i)
                all_tokens.append(Token(token_type, current_number, token_pos))
                current_number = ""
            if char == '\n':
                token_type = token_types[5]
                token_pos = (i, i)
                all_tokens.append(Token(token_type, 'null', token_pos))
        elif char in '+-*/':
            all_tokens.append(Token(token_types[1], char, (i, i+1)))
        elif char == '=':
            all_tokens.append(Token(token_types[6], 'null', (i, i+1)))
        elif char.isdigit():
            current_number += char
        elif char in string.ascii_letters:
            current_string += char
        elif char in '\'"':
            if current_string:
                token_pos = (i - len(current_string), i)
                all_tokens.append(Token(token_types[4], current_string, token_pos))
                current_string = ""
            all_tokens.append(Token(token_types[2], char, (i, i+1)))
        else:
            raise TNLUnidentifiedToken(f"{char}")

    if current_string:
        token_pos = (i - len(current_string) + 1, i + 1)
        all_tokens.append(Token(token_types[4], current_string, token_pos))
    elif current_number:
        token_pos = (i - len(current_number) + 1, i + 1)
        all_tokens.append(Token(token_types[3], current_number, token_pos))
    if all_tokens and all_tokens[-1].type != 'endstatement':
        all_tokens.append(Token(token_types[5], 'null', (i+1, i+1)))

    return all_tokens

File: c++/lexer.cpp
#include "error.cpp"
#include "objects.cpp"
#include <map>
#include <string>
#include <vector>

std::vector<std::string> token_types = {
    "datatype", "operator",     "punctuator", "number",
    "string",   "endstatement", "assignment", "rawstring"};

std::vector<std::string> data_types = {"int", "str"};

bool isNumber(const std::string &s) {
    for (char const &ch : s) {
        if (std::isdigit(ch) == 0)
            return false;
    }
    return true;
}

std::vector<Token> tokenize(std::string &code) {

    std::vector<Token> all_tokens;
    std::vector<char> current_string;
    std::vector<char> current_number;

    bool punct = false;

    for (int k = 0; k < code.length(); k++) {
        std::string i(1, code[k]);
        if (i == " " or i == "\n") {

            for (auto j : data_types) {
                if (std::string(current_string.begin(), current_string.end()) ==
                    j) {
                    all_tokens.push_back(
                        Token("datatype", std::string(current_string.begin(),
                                                      current_string.end())));
                    current_string.clear();
                    continue;
                }
            }
            if (current_string.size() != 0) {
                all_tokens.push_back(
                    Token("string", std::string(current_string.begin(),
                                                current_string.end())));
                current_string.clear();
            }
            if (current_number.size() != 0) {
                all_tokens.push_back(
                    Token("number", std::string(current_number.begin(),
                                                current_number.end())));
                current_number.clear();
            }
            if (i == "\n") {
                all_tokens.push_back(Token("endstatement", "null"));
                current_string.clear();
                current_number.clear();
            }
            punct = false;
            continue;
        }

        if (i == "+" or i == "-" or i == "*" or i == "/") {
            all_tokens.push_back(Token("operator", i));
            continue;
        }

        if (i == "=") {
            all_tokens.push_back(Token("assignment", "null"));
            continue;
        }

        if (isNumber(i)) {
            current_number.push_back(*i.data());
            continue;
        }

        if (std::isalpha(*i.data())) {
            current_string.push_back(*i.data());
            continue;
        }

        if (i == "'" || i == "\"") {
            if (!punct) {
                punct = true;
                all_tokens.push_back(Token("punctuator", i));
                continue;
            }
            if (punct) {
                punct = false;
                all_tokens.push_back(
                    Token("string", std::string(current_string.begin(),
                                                current_string.end())));
                current_string.clear();
                all_tokens.push_back(Token("punctuator", i));
                continue;
            }
        }

        if (true) {
            throw TNLUnidentifiedToken("\n\nUnidentified token " + i + ".");
        }
    }

    if (current_string.size() != 0) {
        all_tokens.push_back(
            Token("string",
                  std::string(current_string.begin(), current_string.end())));
        current_string.clear();
    }
    if (current_number.size() != 0) {
        all_tokens.push_back(
            Token("number",
                  std::string(current_number.begin(), current_number.end())));
        current_number.clear();
    }
    if (all_tokens[all_tokens.size() - 1].gettype() != "endstatement") {
        all_tokens.push_back(Token("endstatement", "null"));
    }

    return all_tokens;
}